src/main/java/org/araymond/joal/core/ttorrent/client/Client.java

```diff
@@ -61,161 +71,169 @@ void setAnnouncerExecutor(final AnnouncerExecutor announcerExecutor) {
 
     @Override
     public void stop() {
+        Lock lock = this.lock.writeLock();
         try {
-            this.lock.writeLock().lock();
+            lock.lock();
             this.stop = true;
             this.torrentFileProvider.unRegisterListener(this);
             if (this.thread != null) {
                 this.thread.interrupt();
                 try {
                     this.thread.join();
                 } catch (final InterruptedException ignored) {
+                } finally {
+                    this.thread = null;
                 }
-                this.thread = null;
             }
             this.delayQueue.drainAll().stream()
-                    .filter(req -> req.getEvent() != RequestEvent.STARTED)
-                    .map(AnnounceRequest::getAnnouncer)
-                    .map(AnnounceRequest::createStop)
+                    .filter(req -> req.getEvent() != RequestEvent.STARTED)  // no need to generate 'stopped' request if the 'started' req was still waiting in queue
+                    .map(AnnounceRequest::toStop)
                     .forEach(this.announcerExecutor::execute);
 
             this.announcerExecutor.awaitForRunningTasks();
         } finally {
-            this.lock.writeLock().unlock();
+            lock.unlock();
         }
     }
 
-    public void onTooManyFailedInARaw(final Announcer announcer) {
-        if (this.stop) {
-            this.currentlySeedingAnnouncer.remove(announcer);
-            return;
-        }
-
+    public void onTooManyFailedInARow(final Announcer announcer) {
+        Lock lock = this.lock.writeLock();
         try {
-            this.lock.writeLock().lock();
-            this.currentlySeedingAnnouncer.remove(announcer); // Remove from announcers list asap, otherwise the deletion will trigger a announce stop event.
-            this.torrentFileProvider.moveToArchiveFolder(announcer.getTorrentInfoHash());
-            this.addTorrent();
+            lock.lock();
+            this.currentlySeedingAnnouncers.remove(announcer);  // Remove from announcers list asap, otherwise the deletion will trigger an announce stop event.
+            if (!this.stop) {
+                this.torrentFileProvider.moveToArchiveFolder(announcer.getTorrentInfoHash());
+                this.addTorrentFromDirectory();
+            }
         } catch (final NoMoreTorrentsFileAvailableException ignored) {
         } finally {
-            this.lock.writeLock().unlock();
+            lock.unlock();
         }
     }
 
     public void onNoMorePeers(final InfoHash infoHash) {
-        if (!this.appConfiguration.isKeepTorrentWithZeroLeechers()) {
+        if (!this.appConfig.isKeepTorrentWithZeroLeechers()) {
             this.torrentFileProvider.moveToArchiveFolder(infoHash);
         }
     }
 
     public void onTorrentHasStopped(final Announcer stoppedAnnouncer) {
-        if (this.stop) {
-            this.currentlySeedingAnnouncer.remove(stoppedAnnouncer);
-            return;
-        }
```

Comment:
hum i don't think so.
With the following function:

    public void onTorrentHasStopped(final Announcer stoppedAnnouncer) {
        if (this.stop) {
            this.currentlySeedingAnnouncer.remove(stoppedAnnouncer);
            return;
        }
        try {
            this.lock.writeLock().lock();

            this.addTorrent();
        } catch (final NoMoreTorrentsFileAvailableException ignored) {
        } finally {
            this.currentlySeedingAnnouncer.remove(stoppedAnnouncer);
            this.lock.writeLock().unlock();
        }
    }

if this.stop == true. The `finally` is not executed since we have a `return` in the if scope.

So the code you propose has not the same effect as the original one

Reply:
Same as above, the proposed code differ from the original

Reply:
> The finally is not executed

Which is why there's duplication of ` this.currentlySeedingAnnouncer.remove(stoppedAnnouncer);` -- ie the block _effectively_ is executed.

Reply:
Ok i see what you mean then.

There is also another problem that led me to this implemantation.

`onTorrentHasStopped()` can be called while `stop()` has not exited yet. My implementation would not cause deadlock in such cases. The new one will.
That's an awfull implementation in the first place 😆 but it need to stay like that.

The `Client` class is a pain to refactor because it has not been well design in the first place. Any change is most likely going to end in a deadlock here 😞

Reply:
You're right. A corner case, but very much a valid one. I'll revert this change in those two locations.